smpp/server.go

```diff
@@ -215,12 +223,31 @@ func (srv *server) Handle(id pdu.ID, h RequestHandlerFunc) {
 	srv.m[id] = h
 }
 
-// auth authenticate new clients.
-func (srv *server) auth(c *conn) error {
+func (srv *server) HandleAuth(id pdu.ID, h AuthRequestHandlerFunc) {
+	srv.a[id] = h
+}
+
+func (srv *server) auth(c *conn, s Session) error {
 	p, err := c.Read()
 	if err != nil {
 		return err
 	}
+
+	// Read the bind PDU and if there are any handlers set, use them,
+	// if not perform the default auth
+	if h, ok := srv.a[p.Header().ID]; ok {
+		return h(s, p)
+	}
+
+	resp, err := srv.defaultAuth(p)
```

Comment:
I see the signature of the defaultAuth is :

```
func (srv *server) defaultAuth(p pdu.Body) (pdu.Body, error)```

Why handler has a different signature, it could return the same thing ?